File: R/TOSTone.R
#' TOST function for a one-sample t-test (Cohen's d)
#' @param m mean
#' @param mu value to compare against
#' @param sd standard deviation
#' @param n sample size
#' @param low_eqbound_d lower equivalence bounds (e.g., -0.5) expressed in standardized mean difference (Cohen's d)
#' @param high_eqbound_d upper equivalence bounds (e.g., 0.5) expressed in standardized mean difference (Cohen's d)
#' @param alpha alpha level (default = 0.05)
#' @param plot set whether results should be plotted (plot = TRUE) or not (plot = FALSE) - defaults to TRUE
#' @param verbose logical variable indicating whether text output should be generated (verbose = TRUE) or not (verbose = FALSE) - default to TRUE
#' @return Returns TOST t-value 1, TOST p-value 1, TOST t-value 2, TOST p-value 2, degrees of freedom, low equivalence bound, high equivalence bound, Lower limit confidence interval TOST, Upper limit confidence interval TOST
#' @examples
#' ## Test observed mean of 0.54 and standard deviation of 1.2 in sample of 100 participants
#' ## against 0.5 given equivalence bounds of Cohen's d = -0.3 and 0.3, with an alpha = 0.05.
#' TOSTone(m=0.54,mu=0.5,sd=1.2,n=100,low_eqbound_d=-0.3, high_eqbound_d=0.3, alpha=0.05)
#' @importFrom stats pnorm pt qnorm qt
#' @importFrom graphics abline plot points segments title
#' @export
#'

TOSTone<-function(m,mu,sd,n,low_eqbound_d, high_eqbound_d, alpha, plot = TRUE, verbose = TRUE){
  message("Note: this function is defunct. Please use tsum_TOST instead")
  if(missing(alpha)) {
    alpha<-0.05
  }
  if(low_eqbound_d >= high_eqbound_d) warning("The lower bound is equal to or larger than the upper bound. Check the plot and output to see if the bounds are specified as you intended.")
  if(n < 2) stop("The sample size should be larger than 1.")
  if(1 <= alpha | alpha <= 0) stop("The alpha level should be a positive value between 0 and 1.")
  if(sd <= 0) stop("The standard deviation should be a positive value.")
  # Calculate TOST, t-test, 90% CIs and 95% CIs
  low_eqbound<-low_eqbound_d*sd
  high_eqbound<-high_eqbound_d*sd
  degree_f<-n-1
  t1<-(m-mu-low_eqbound)/(sd/sqrt(n))# t-test
  p1<-pt(t1, degree_f, lower.tail=FALSE)
  t2<-(m-mu-high_eqbound)/(sd/sqrt(n)) #t-test
  p2<-pt(t2, degree_f, lower.tail=TRUE)
  t<-(m-mu)/(sd/sqrt(n))
  pttest<-2*pt(-abs(t), df=degree_f)
  LL90<-(m-mu)-qt(1-alpha, degree_f)*(sd/sqrt(n))
  UL90<-(m-mu)+qt(1-alpha, degree_f)*(sd/sqrt(n))
  LL95<-(m-mu)-qt(1-(alpha/2), degree_f)*(sd/sqrt(n))
  UL95<-(m-mu)+qt(1-(alpha/2), degree_f)*(sd/sqrt(n))
  ptost<-max(p1,p2) #Get highest p-value for summary TOST result
  ttost<-ifelse(abs(t1) < abs(t2), t1, t2) #Get lowest t-value for summary TOST result
  dif<-(m-mu)
  testoutcome<-ifelse(pttest<alpha,"significant","non-significant")
  TOSToutcome<-ifelse(ptost<alpha,"significant","non-significant")

  # Plot results
  if (plot == TRUE) {
  plot(NA, ylim=c(0,1), xlim=c(min(LL95,low_eqbound,dif)-max(UL95-LL95, high_eqbound-low_eqbound,dif)/10, max(UL95,high_eqbound,dif)+max(UL95-LL95, high_eqbound-low_eqbound, dif)/10), bty="l", yaxt="n", ylab="",xlab="Mean Difference")
  points(x=dif, y=0.5, pch=15, cex=2)
  abline(v=high_eqbound, lty=2)
  abline(v=low_eqbound, lty=2)
  abline(v=0, lty=2, col="grey")
  segments(LL90,0.5,UL90,0.5, lwd=3)
  segments(LL95,0.5,UL95,0.5, lwd=1)
  title(main=paste("Equivalence bounds ",round(low_eqbound,digits=3)," and ",round(high_eqbound,digits=3),"\nMean difference = ",round(dif,digits=3)," \n TOST: ", 100*(1-alpha*2),"% CI [",round(LL90,digits=3),";",round(UL90,digits=3),"] ", TOSToutcome," \n NHST: ", 100*(1-alpha),"% CI [",round(LL95,digits=3),";",round(UL95,digits=3),"] ", testoutcome,sep=""), cex.main=1)
  }

  if(missing(verbose)) {
    verbose <- TRUE
  }
  if(verbose == TRUE){
    cat("TOST results:\n")
    cat("t-value lower bound:",format(t1, digits = 3, nsmall = 2, scientific = FALSE),"\tp-value lower bound:",format(p1, digits = 1, nsmall = 3, scientific = FALSE))
    cat("\n")
    cat("t-value upper bound:",format(t2, digits = 3, nsmall = 2, scientific = FALSE),"\tp-value upper bound:",format(p2, digits = 1, nsmall = 3, scientific = FALSE))
    cat("\n")
    cat("degrees of freedom :",round(degree_f, digits = 2))
    cat("\n\n")
    cat("Equivalence bounds (Cohen's d):")
    cat("\n")
    cat("low eqbound:", paste0(round(low_eqbound_d, digits = 4)),"\nhigh eqbound:",paste0(round(high_eqbound_d, digits = 4)))
    cat("\n\n")
    cat("Equivalence bounds (raw scores):")
    cat("\n")
    cat("low eqbound:", paste0(round(low_eqbound, digits = 4)),"\nhigh eqbound:",paste0(round(high_eqbound, digits = 4)))
    cat("\n\n")
    cat("TOST confidence interval:")
    cat("\n")
    cat("lower bound ",100*(1-alpha*2),"% CI: ", paste0(round(LL90, digits = 3)),"\nupper bound ",100*(1-alpha*2),"% CI:  ",paste0(round(UL90,digits = 3)), sep = "")
    cat("\n\n")
    cat("NHST confidence interval:")
    cat("\n")
    cat("lower bound ",100*(1-alpha),"% CI: ", paste0(round(LL95, digits = 3)),"\nupper bound ",100*(1-alpha),"% CI:  ",paste0(round(UL95,digits = 3)), sep = "")
    cat("\n\n")
    cat("Equivalence Test Result:\n")
    message(cat("The equivalence test was ",TOSToutcome,", t(",round(degree_f, digits=2),") = ",format(ttost, digits = 3, nsmall = 3, scientific = FALSE),", p = ",format(ptost, digits = 3, nsmall = 3, scientific = FALSE),", given equivalence bounds of ",format(low_eqbound, digits = 3, nsmall = 3, scientific = FALSE)," and ",format(high_eqbound, digits = 3, nsmall = 3, scientific = FALSE)," (on a raw scale) and an alpha of ",alpha,".",sep=""))
    cat("\n")
    cat("Null Hypothesis Test Result:\n")
    message(cat("The null hypothesis test was ",testoutcome,", t(",round(degree_f, digits=2),") = ",format(t, digits = 3, nsmall = 3, scientific = FALSE),", p = ",format(pttest, digits = 3, nsmall = 3, scientific = FALSE),", given an alpha of ",alpha,".",sep=""))
    if(pttest <= alpha && ptost <= alpha){
      combined_outcome <- "statistically different from zero and statistically equivalent to zero"
    }
    if(pttest < alpha && ptost > alpha){
      combined_outcome <- "statistically different from zero and statistically not equivalent to zero"
    }
    if(pttest > alpha && ptost <= alpha){
      combined_outcome <- "statistically not different from zero and statistically equivalent to zero"
    }
    if(pttest > alpha && ptost > alpha){
      combined_outcome <- "statistically not different from zero and statistically not equivalent to zero"
    }
    cat("\n")
    message(cat("Based on the equivalence test and the null-hypothesis test combined, we can conclude that the observed effect is ",combined_outcome,".",sep=""))
  }
  # Return results in list()
  invisible(list(diff=dif,TOST_t1=t1,TOST_p1=p1,TOST_t2=t2,TOST_p2=p2, TOST_df=degree_f,alpha=alpha,low_eqbound=low_eqbound,high_eqbound=high_eqbound,low_eqbound_d=low_eqbound_d,high_eqbound_d=high_eqbound_d, LL_CI_TOST=LL90,UL_CI_TOST=UL90, LL_CI_TTEST=LL95, UL_CI_TTEST=UL95,NHST_t = t, NHST_p = pttest))
}

File: pyTOSTER/tost_one.py
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt

def tost_one(m, mu, sd, n, low_eqbound_d, high_eqbound_d, alpha=0.05, plot=True, verbose=True):
    """
    TOST function for a one-sample t-test (Cohen's d).

    Parameters
    ----------
    m : float
        mean.
    mu : float
        value to compare against.
    sd : float
        standard deviation.
    n : int
        sample size.
    low_eqbound_d : float
        lower equivalence bounds (e.g., -0.5) expressed in standardized mean difference (Cohen's d).
    high_eqbound_d : float
        upper equivalence bounds (e.g., 0.5) expressed in standardized mean difference (Cohen's d).
    alpha : float, optional
        alpha level (default = 0.05).
    plot : bool, optional
        set whether results should be plotted (default is True).
    verbose : bool, optional
        logical variable indicating whether text output should be generated (default is True).

    Returns
    -------
    dict
        A dictionary containing the TOST results.
    """

    if low_eqbound_d >= high_eqbound_d:
        print("Warning: The lower bound is equal to or larger than the upper bound.")
    if n < 2:
        raise ValueError("The sample size should be larger than 1.")
    if not (0 < alpha < 1):
        raise ValueError("The alpha level should be a positive value between 0 and 1.")
    if sd <= 0:
        raise ValueError("The standard deviation should be a positive value.")

    low_eqbound = low_eqbound_d * sd
    high_eqbound = high_eqbound_d * sd

    return tost_one_raw(m, mu, sd, n, low_eqbound, high_eqbound, alpha, plot, verbose,
                        low_eqbound_d=low_eqbound_d, high_eqbound_d=high_eqbound_d)


def tost_one_raw(m, mu, sd, n, low_eqbound, high_eqbound, alpha=0.05, plot=True, verbose=True, low_eqbound_d=None, high_eqbound_d=None):
    """
    TOST function for a one-sample t-test (raw scores).
    """

    if low_eqbound >= high_eqbound:
        print("Warning: The lower bound is equal to or larger than the upper bound.")
    if n < 2:
        raise ValueError("The sample size should be larger than 1.")
    if not (0 < alpha < 1):
        raise ValueError("The alpha level should be a positive value between 0 and 1.")
    if sd <= 0:
        raise ValueError("The standard deviation should be a positive value.")

    degree_f = n - 1

    t1 = (m - mu - low_eqbound) / (sd / np.sqrt(n))
    p1 = 1 - stats.t.cdf(t1, df=degree_f)
    t2 = (m - mu - high_eqbound) / (sd / np.sqrt(n))
    p2 = stats.t.cdf(t2, df=degree_f)

    t = (m - mu) / (sd / np.sqrt(n))
    pttest = 2 * (1 - stats.t.cdf(abs(t), df=degree_f))

    ll90 = (m - mu) - stats.t.ppf(1 - alpha, df=degree_f) * (sd / np.sqrt(n))
    ul90 = (m - mu) + stats.t.ppf(1 - alpha, df=degree_f) * (sd / np.sqrt(n))
    ll95 = (m - mu) - stats.t.ppf(1 - (alpha / 2), df=degree_f) * (sd / np.sqrt(n))
    ul95 = (m - mu) + stats.t.ppf(1 - (alpha / 2), df=degree_f) * (sd / np.sqrt(n))

    ptost = max(p1, p2)
    ttost = t1 if abs(t1) < abs(t2) else t2
    diff = m - mu

    test_outcome = "significant" if pttest < alpha else "non-significant"
    tost_outcome = "significant" if ptost < alpha else "non-significant"

    if plot:
        plt.figure()
        plt.ylim(0, 1)
        plot_min = min(ll95, low_eqbound, diff) - max(ul95 - ll95, high_eqbound - low_eqbound, diff) / 10
        plot_max = max(ul95, high_eqbound, diff) + max(ul95 - ll95, high_eqbound - low_eqbound, diff) / 10
        plt.xlim(plot_min, plot_max)

        plt.yticks([])
        plt.xlabel("Mean Difference")
        title = (f"Equivalence bounds {low_eqbound:.3f} and {high_eqbound:.3f}\\n"
                 f"Mean difference = {diff:.3f}\\n"
                 f"TOST: {100*(1-alpha*2):.0f}% CI [{ll90:.3f}; {ul90:.3f}] {tost_outcome}\\n"
                 f"NHST: {100*(1-alpha):.0f}% CI [{ll95:.3f}; {ul95:.3f}] {test_outcome}")
        plt.title(title.replace('\\n', '\n'))

        plt.plot(diff, 0.5, 's', markersize=8, color='black')
        plt.axvline(high_eqbound, linestyle='--', color='black')
        plt.axvline(low_eqbound, linestyle='--', color='black')
        plt.axvline(0, linestyle='--', color='grey')

        plt.hlines(0.5, ll90, ul90, linewidth=3, color='black')
        plt.hlines(0.5, ll95, ul95, linewidth=1, color='black')

        plt.show()

    if verbose:
        print("TOST results:")
        print(f"t-value lower bound: {t1:8.2f}\\tp-value lower bound: {p1:.3f}")
        print(f"t-value upper bound: {t2:8.2f}\\tp-value upper bound: {p2:.3f}")
        print(f"degrees of freedom : {degree_f:.2f}")
        print("\\n")
        if low_eqbound_d is not None and high_eqbound_d is not None:
            print("Equivalence bounds (Cohen's d):")
            print(f"low eqbound: {low_eqbound_d:.4f}\\nhigh eqbound: {high_eqbound_d:.4f}")
            print("\\n")
        print("Equivalence bounds (raw scores):")
        print(f"low eqbound: {low_eqbound:.4f}\\nhigh eqbound: {high_eqbound:.4f}")
        print("\\n")
        print("TOST confidence interval:")
        print(f"lower bound {100*(1-alpha*2):.0f}% CI: {ll90:.3f}\\nupper bound {100*(1-alpha*2):.0f}% CI:  {ul90:.3f}")
        print("\\n")
        print("NHST confidence interval:")
        print(f"lower bound {100*(1-alpha):.0f}% CI: {ll95:.3f}\\nupper bound {100*(1-alpha):.0f}% CI:  {ul95:.3f}")
        print("\\n")
        print("Equivalence Test Result:")
        print(f"The equivalence test was {tost_outcome}, t({degree_f:.2f}) = {ttost:.3f}, p = {ptost:.3f}, given equivalence bounds of {low_eqbound:.3f} and {high_eqbound:.3f} (on a raw scale) and an alpha of {alpha}.")
        print("\\n")
        print("Null Hypothesis Test Result:")
        print(f"The null hypothesis test was {test_outcome}, t({degree_f:.2f}) = {t:.3f}, p = {pttest:.3f}, given an alpha of {alpha}.")

        if pttest <= alpha and ptost <= alpha:
            combined_outcome = (f"NHST: reject null significance hypothesis that the effect is equal to 0 \\n"
                                f"TOST: reject null equivalence hypothesis")
        elif pttest < alpha and ptost > alpha:
            combined_outcome = (f"NHST: reject null significance hypothesis that the effect is equal to 0 \\n"
                                f"TOST: don't reject null equivalence hypothesis")
        elif pttest > alpha and ptost <= alpha:
            combined_outcome = (f"NHST: don't reject null significance hypothesis that the effect is equal to 0 \\n"
                                f"TOST: reject null equivalence hypothesis")
        else: # pttest > alpha and ptost > alpha
            combined_outcome = (f"NHST: don't reject null significance hypothesis that the effect is equal to 0 \\n"
                                  f"TOST: don't reject null equivalence hypothesis")

        print("\\n")
        print(combined_outcome)

    results = {
        'diff': diff,
        'TOST_t1': t1,
        'TOST_p1': p1,
        'TOST_t2': t2,
        'TOST_p2': p2,
        'TOST_df': degree_f,
        'alpha': alpha,
        'low_eqbound': low_eqbound,
        'high_eqbound': high_eqbound,
        'LL_CI_TOST': ll90,
        'UL_CI_TOST': ul90,
        'LL_CI_TTEST': ll95,
        'UL_CI_TTEST': ul95,
        'NHST_t': t,
        'NHST_p': pttest
    }
    if low_eqbound_d is not None:
        results['low_eqbound_d'] = low_eqbound_d
    if high_eqbound_d is not None:
        results['high_eqbound_d'] = high_eqbound_d

    return results
